test/unittest/unittest_nntrainer_tensor_pool.cpp

```diff
@@ -0,0 +1,364 @@
+
+  EXPECT_NO_THROW(pool.requestTensor(nntrainer::TensorDim({1}), {},
+                                     nntrainer::TensorLifespan::ZERO_LIFESPAN,
+                                     "abc"));
+
+  EXPECT_THROW(pool.requestPreallocatedTensor(
+                 nntrainer::TensorDim({1}), {},
+                 nntrainer::TensorLifespan::ZERO_LIFESPAN, "not_exist"),
+               std::invalid_argument);
+}
+
+/**
+ * @brief request already allocated tensor
+ */
+TEST(TensorPool, request_mem_08_p) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1, *t2;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {},
+                            nntrainer::TensorLifespan::ZERO_LIFESPAN, "abc"));
+  EXPECT_NE(t1, nullptr);
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_NO_THROW(t2 = pool.requestPreallocatedTensor(
+                    nntrainer::TensorDim({1}), {},
+                    nntrainer::TensorLifespan::MAX_LIFESPAN, "abc"));
+  EXPECT_NE(t2, nullptr);
+  EXPECT_FALSE(t2->isAllocated());
+
+  EXPECT_EQ(t1, t2);
+
+  EXPECT_NO_THROW(t2 = pool.requestPreallocatedTensor(
+                    nntrainer::TensorDim({1}), {},
+                    nntrainer::TensorLifespan::MAX_LIFESPAN, "abc"));
+  EXPECT_NE(t2, nullptr);
+  EXPECT_FALSE(t2->isAllocated());
+
+  EXPECT_EQ(t1, t2);
+}
+
+/**
+ * @brief set batch
+ */
+TEST(TensorPool, set_batch_01_p) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {},
+                            nntrainer::TensorLifespan::ZERO_LIFESPAN, "abc"));
+  EXPECT_NE(t1, nullptr);
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_EQ(t1->batch(), 1);
+  EXPECT_NO_THROW(pool.setBatchSize("abc", 10));
+  EXPECT_EQ(t1->batch(), 10);
+}
+
+/**
+ * @brief set batch
+ */
+TEST(TensorPool, set_batch_02_n) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {},
+                            nntrainer::TensorLifespan::ZERO_LIFESPAN, "abc"));
+  EXPECT_NE(t1, nullptr);
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_THROW(pool.setBatchSize("not_exist", 10), std::invalid_argument);
+  EXPECT_EQ(t1->batch(), 1);
+}
+
+/**
+ * @brief zero size pool as no usage
+ */
+TEST(TensorPool, finalize_01_p) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1, *t2;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {},
+                            nntrainer::TensorLifespan::ZERO_LIFESPAN, "abc1"));
+  EXPECT_NE(t1, nullptr);
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_NO_THROW(
+    t2 = pool.requestTensor(nntrainer::TensorDim({1}), {},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc2"));
+  EXPECT_NE(t2, nullptr);
+  EXPECT_FALSE(t2->isAllocated());
+
+  EXPECT_NE(t1, t2);
+
+  EXPECT_NO_THROW(pool.finalize(nntrainer::BasicPlanner(), 0, 2));
+  EXPECT_EQ(pool.minMemoryRequirement(), 0);
```

Comment:
is this zero?

Reply:
Yes, it is zero because the tensors, although with a valid lifespan, have no usage. As they aren't expected to be ever used, they aren't allocated any memory.

Reply:
I see